File: foo.js
import * as fs from 'fs';
const bytes = fs.readFileSync('./foo.wasm');

const print_bool = (value, dataView) => {
  if (value & 1) {
    switch (value >> 1) {
      case 0: process.stdout.write('true');
      case 1: process.stdout.write('false');
    }
  }
};

const importObject = {
  env: {
    write_int(value) {
      process.stdout.write(value.toString());
    },
    write_char(value) {
      process.stdout.write(String.fromCharCode(value));
    },
  },
};

(async () => {
  const object = await WebAssembly.instantiate(
    new Uint8Array(bytes), importObject,
  );

  try {
    const start = Date.now();
    object.instance.exports.main_function();
    const stop = Date.now();

    const memory = object.instance.exports.memory;
    const dataView = new DataView(memory.buffer);
    const res_value = object.instance.exports.result.value;
    process.stdout.write('====> ');
    print_bool(res_value, dataView);

    const bytes = object.instance.exports.bytes_used.value;
    console.log(`\n====> used ${bytes} bytes of memory, took ${(stop - start)} (Node.js) ms.`);
  } catch (error) {
    console.log(error);
    process.exit(1);
  }
})();

File: foo.py
from wasmtime import Store, Module, Instance, Func, FuncType, ValType
import sys
import time

# Almost all operations in wasmtime require a contextual "store" argument to be
# shared amongst objects
store = Store()

# Here we can compile a `Module` which is then ready for instantiation
# afterwards
module = Module.from_file(store.engine, './foo.wasm')


# Our module needs two imports
def print_int(n):
    sys.stdout.write(str(n))


def print_char(n):
    sys.stdout.write(str(chr(n)))


print_char_stdout = Func(store, FuncType([ValType.i32()], []), print_char)
print_int_stdout = Func(store, FuncType([ValType.i32()], []), print_int)

# instantiate module
instance = Instance(store, module, [print_char_stdout, print_int_stdout])

# run main
start = time.time()
instance.exports(store)["main_function"](store)
end = time.time()

result = instance.exports(store)["result"].value(store)
sys.stdout.write("====> ")
instance.exports(store)["pretty_print_constructor"](store, result)
print()

bytes_used = instance.exports(store)["bytes_used"].value(store)
print(f"====> Used {bytes_used} bytes of memory, took {round(1000 * (end - start), 2)} ms (wasmtime).")
print()
